### ISSUE TITLE
Navigation api

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/e2e-testing#36]
Since we moved #5 to a later milestone, I think I have to create a new issue just to clarify how we'll do navigation.

**Comment by inf3rno** [comment]
I just realized that the age of XHR is over, and there is a fetch api. https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API I think our navigation could learn from that api too. I need to check the phantomjs and selenium apis too. #35

**Comment by inf3rno** [comment]
I think we should build our low level api on the common events people use: DOMContentLoaded and onload. In #26 we tested page loading from the parent window perspective:

```js
LOG: 'loading', 'http://localhost:3333/'
LOG: 'document.addEventListener:readystatechange', 'http://localhost:3333/', 'interactive'
LOG: 'document.addEventListener:DOMContentLoaded', 'http://localhost:3333/'
LOG: 'win.addEventListener:DOMContentLoaded', 'http://localhost:3333/'
LOG: 'document.addEventListener:readystatechange', 'http://localhost:3333/', 'complete'
LOG: 'win.onload', 'http://localhost:3333/'
LOG: 'win.addEventListener:load', 'http://localhost:3333/'
LOG: 'timeout 1000', 'complete', 'http://localhost:3333/'
```

I think we should load similar testing code in the client too, so we will know more about the loading process and we might answer the #34 too.

**Comment by inf3rno** [comment]
The logs:

```
LOG: 1534379227173, 'parent', 'document.addEventListener:readystatechange', 'http://localhost:9876/base/test/karma.static/helloworld.html', 'interactive'
LOG: 1534379227174, 'parent', 'document.addEventListener:DOMContentLoaded', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: 1534379227174, 'parent', 'win.addEventListener:DOMContentLoaded', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: 1534379227175, 'parent', 'document.addEventListener:readystatechange', 'http://localhost:9876/base/test/karma.static/helloworld.html', 'complete'
LOG: 1534379227175, 'parent', 'win.onload', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: 1534379227175, 'parent', 'win.addEventListener:load', 'http://localhost:9876/base/test/karma.static/helloworld.html'

LOG: 1534379228186, 'parent', 'document.addEventListener:readystatechange', 'http://localhost:3333/', 'interactive'
LOG: 1534379228186, 'parent', 'document.addEventListener:DOMContentLoaded', 'http://localhost:3333/'
LOG: 1534379228186, 'parent', 'win.addEventListener:DOMContentLoaded', 'http://localhost:3333/'
LOG: 1534379228186, 'parent', 'document.addEventListener:readystatechange', 'http://localhost:3333/', 'complete'
LOG: 1534379228187, 'parent', 'win.onload', 'http://localhost:3333/'
LOG: 1534379228187, 'parent', 'win.addEventListener:load', 'http://localhost:3333/'

LOG: 1534379229603, 'parent', 'document.addEventListener:readystatechange', 'https://www.google.com/', 'interactive'
LOG: 1534379229604, 'parent', 'document.addEventListener:DOMContentLoaded', 'https://www.google.com/'
LOG: 1534379229611, 'parent', 'win.addEventListener:DOMContentLoaded', 'https://www.google.com/'
LOG: 1534379230018, 'parent', 'document.addEventListener:readystatechange', 'https://www.google.com/', 'complete'
LOG: 1534379230020, 'parent', 'win.addEventListener:load', 'https://www.google.com/'
```

**Comment by inf3rno** [comment]
By google.com the `win.onload` is overridden by the site, so better not to use that and stick to the other listeners. It is interesting that by google.com all of them have different timestamp. I think we should use microsecond resolution by logging that. Now add the same code to a local server page.

**Comment by inf3rno** [comment]
ms resolution is enough:

```
LOG: '137.300', 'parent', 'document.addEventListener:readystatechange', 'http://localhost:9876/base/test/karma.static/helloworld.html', 'interactive'
LOG: '138.400', 'parent', 'document.addEventListener:DOMContentLoaded', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '138.500', 'parent', 'win.addEventListener:DOMContentLoaded', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '138.700', 'parent', 'document.addEventListener:readystatechange', 'http://localhost:9876/base/test/karma.static/helloworld.html', 'complete'
LOG: '138.900', 'parent', 'win.onload', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '139.000', 'parent', 'win.addEventListener:load', 'http://localhost:9876/base/test/karma.static/helloworld.html'
```

**Comment by inf3rno** [comment]
Okay, we can get the URI setting time: `win.performance.timing.navigationStart` and the time elapsed since then: `win.performance.now()`. This is different for every window and it is in msecs, not in secs. If we add the 2 together, we got a high res timestamp in msecs. I don't think we need a timestamp, it is better to use the elapsedTime and the window object of the child window can be reached from both the parent and the child.

**Comment by inf3rno** [comment]
I think we should write a self tests for event order after we are finished with the API. In theory we can add our events to several points: 

```
<-- point 1 - after: loading / before: interactive
LOG: '137.300', 'parent', 'document.addEventListener:readystatechange', 'http://localhost:9876/base/test/karma.static/helloworld.html', 'interactive'
LOG: '138.400', 'parent', 'document.addEventListener:DOMContentLoaded', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '138.500', 'parent', 'win.addEventListener:DOMContentLoaded', 'http://localhost:9876/base/test/karma.static/helloworld.html'
<-- point 2 - interactive
tick
<-- point 3 - after: interactive / before: complete
LOG: '138.700', 'parent', 'document.addEventListener:readystatechange', 'http://localhost:9876/base/test/karma.static/helloworld.html', 'complete'
LOG: '138.900', 'parent', 'win.onload', 'http://localhost:9876/base/test/karma.static/helloworld.html'
LOG: '139.000', 'parent', 'win.addEventListener:load', 'http://localhost:9876/base/test/karma.static/helloworld.html'
<-- point 4 - complete
tick
<-- point 5 - after: complete
```

We need to decide whether we need all of these.

 - after: loading / before: interactive
 - interactive
 - after: interactive / before : complete
 - complete
 - after: complete

These are 5+2 events and we can use different naming for them, e.g. navigation, document created, uri set, dom content loaded, loaded, load, etc.. I think it is hard to reliably name them if we don't use the upper before and after events. Normally what do we need? DOMContentLoaded, onload, I think that's all... For the tests it is good to have an after: onload handler, so we can be sure that the site loaded all necessary scripts and it is responsive. But we don't need a sugar syntax for that. We can use onload + nexttick instead. If so, then we need only loading, interactive and complete and we can mock loading with after:loading, because nobody will know the difference, except if a script is loaded before unload + nexttick. We need to check for that, but in #34 we do exactly that experiment.

Another option to do something completely different. Instead of readystatechange, which is very long, I would add a lifecycle event to the client. This should contain the readyStates, and other events like "unloading" (beforeunload), "navigating" (unload).

So a page lifecycle would be:

- navigating
- loading
- interactive
- complete
- unloading

I think that is a good start and we could add history events, hash change events, etc. later. I think on the long run everything connected to navigation must be covered here. I like the before and after prefixes too.

**Comment by inf3rno** [comment]
I think every URI change must be part of the API. so onhashchange and onpopstate too
The onpopstate is funny: https://stackoverflow.com/questions/15896434/window-onpopstate-on-page-load According to it Chrome and Opera emits it by page load, while Firefox doesn't. It doesn't really matter because we can't add it to the new document.

**Comment by inf3rno** [comment]
I think the basic api (which does not contain pushState and history changes) is clear now:

These are the events:

 - page.unload/hide
 - navigation.request
 - navigation.response
 - navigation.error
 - page.loading
 - page.interactive
 - page.load/show
 - page.error

These are the operations:

 - windowManager.load(resourceIdentifier) -> {page.load: resolve, navigation.error: reject}
 - windowManager.execute(function (resolve){...}) -> {res: resolve, err: reject}

For now custom event handlers are not supported. You can get the location with execute. For now this should be enough. Later I'll add more features, but this is hard enough to implement, especially the navigation part.

**Comment by inf3rno** [comment]
I think this can be closed.
